src/Slot2Cart.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+    Copyright 2019 Arisotura, Buenia0
+
+    This file is part of melonDS.
+
+    melonDS is free software: you can redistribute it and/or modify it under
+    the terms of the GNU General Public License as published by the Free
+    Software Foundation, either version 3 of the License, or (at your option)
+    any later version.
+
+    melonDS is distributed in the hope that it will be useful, but WITHOUT ANY
+    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with melonDS. If not, see http://www.gnu.org/licenses/.
+*/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "Slot2Cart.h"
+#include "CRC32.h"
+#include "Platform.h"
+
+namespace Slot2Cart_RumblePak
+{
+
+bool RumblePakEnabled = false;
+u16 RumbleState = 0;
+
+u16 ReadRumble(u32 addr)
+{
+    // GBATEK: For detection, AD1 seems to be pulled low when reading from it... (while) the other AD lines are open bus (containing the halfword address)...
+    // GBATEK with bit 6 set is based on empirical data...
+    // This should allow commercial games to properly detect the Rumble Pak.
+    // Credit to rzumer for coming up with this algorithm...
+    
+    u8 lodata = ((addr | 0x40) & 0xFFFD);
```

Comment:
This should be u16. The upper byte doesn't seem to matter anyway, but it matches GBATEK more and if we're returning a 16-bit value that's mostly open bus there's no reason to clear the upper byte.